AWS/aws/event_aggregator.py

```diff
@@ -0,0 +1,169 @@
+import datetime
+import threading
+import time
+from typing import Callable
+from ciso8601 import parse_datetime_as_naive
+
+from pydantic import BaseModel
+
+
+class Fingerprint(BaseModel):
+    condition_func: Callable[[dict], bool]
+    build_hash_str_func: Callable[[dict], str]
+
+
+class Aggregation(BaseModel):
+    start: datetime.datetime
+    end: datetime.datetime
+    count: int
+    event: dict
+
+    def get_aggregated_event(self):
+        aggregated_event = self.event
+        aggregated_event["event"]["start"] = self.start.isoformat()
+        aggregated_event["event"]["end"] = self.start.isoformat()
+        aggregated_event["event"]["count"] = self.count
+        return aggregated_event
+
+
+class EventAggregatorTTLThread(threading.Thread):
+    f_must_stop: bool  # thread will stop if this flag is active
+    ttl: int  # ttl value
+    on_flush_func: Callable[[dict], None]  # function to be call when aggregation is flushed
+
+    def __init__(
+        self,
+        event_aggregator: "EventAggregator",
+        ttl: int,
+        on_flush_func: Callable[[dict], None],
+        delay: float = 10,
+    ):
+        super().__init__()
+        self.event_aggregator = event_aggregator
+        self.f_must_stop = False
+        self.ttl = ttl
+        self.on_flush_func = on_flush_func
+        self.delay = delay
+
+    def stop(self) -> None:
+        self.f_must_stop = True
+
+    def run(self) -> None:
+        try:
+            while not self.f_must_stop:
+                time.sleep(self.delay)
+                for aggregated_event in self.event_aggregator.flush_all_ttl(ttl=self.ttl):
+                    self.on_flush_func(aggregated_event)
+        finally:
+            for aggregated_event in self.event_aggregator.flush_all():
+                self.on_flush_func(aggregated_event)
+
+
+class EventAggregator:
+    ttl_thread: EventAggregatorTTLThread | None
+    lock: threading.Lock
+
+    def __init__(self, aggregation_definitions: dict[str, list[Fingerprint]]):
+        self.aggregation_definitions = aggregation_definitions
+        self.aggregations: dict[str, Aggregation] = dict()
+        self.lock = threading.Lock()
+
+    def start_flush_on_ttl(self, ttl: int, on_flush_func: Callable[[dict], None], delay: float = 10):
```

Comment:
TTL is global and can't be changed depending on the kind of event. 

I'm totally fine with it, but I'm wondering if we can have cases where some aggregations should be forwarded faster than other ?

Reply:
well spotted, I will upgrade the aggregation definition to support custom ttl